src/StableHashMap.mo

```diff
@@ -0,0 +1,243 @@
+/// Mutable hash map (aka Hashtable)
+///
+/// This module defines an imperative hash map (hash table), with a general key and value type.
+///
+/// It is like HashMap, but eschews classes and objects so that its type is `stable`.
+///
+
+import Prim "mo:⛔";
+import P "Prelude";
+import A "Array";
+import Hash "Hash";
+import Iter "Iter";
+import AssocList "AssocList";
+
+module {
+  // Type parameters that we need for all operations:
+  // - K = key type
+  // - V = value type
+
+
+  // key-val list
+  public type KVs<K, V> = AssocList.AssocList<K, V>;
+
+  // representation of (the stable types of) the hash map state.
+  public type HashMap<K, V> = {
+    var table : [var KVs<K, V>];
+    var count : Nat;
+  };
+
+  // representation of hash map including key operations.
+  // unlike HashMap, this type is not stable, but is required for
+  // some operations (keyEq and keyHash are functions).
+  public type HashMap_<K, V> = {
+    keyEq : (K, K) -> Bool;
+    keyHash : K -> Hash.Hash;
+    // to do -- use union type operation to compress this definition
```

Comment:
It's intersection, actually. :)

Reply:
> It's intersection, actually. :)

I perennially invert that operation for records of functions.  Not sure why.  Thanks.

>  impossible to perform relevant operations on an actual stable hash-map ... What would I actually do with a plain `HashMap`?

I admit that the API for that type is limited, but you can still do useful things, arguably.

One could fill a `Buffer` with the contents efficiently, via 
- `size` (to pre-allocate that `Buffer` of the right size) and 
- `entries`, to iterate over the elements.

> ... which I can't even use `get` on?

Well, `get` requires key equality and hashing, so you need to have the non-stable `HashMap_` version to project specific keys, and call those operations.